src/cpp/r/RSexp.cpp

```diff
@@ -689,6 +690,17 @@ bool isPrimitiveEnvironment(SEXP object)
    return TYPEOF(object) == ENVSXP;
 }
 
+bool isUserDefinedDatabase(SEXP object)
+{
+   return OBJECT(object) && Rf_inherits(object, "UserDefinedDatabase");
+}
+
+bool isActiveBinding(SEXP binding) 
```

Comment:
I feel a little skittish about using the SEXP internals directly here, but unfortunately it doesn't seem like there's an API for asking if a binding directly is active...

Reply:
yeah the only api we have available is `R_BindingIsActive()` which has to do the lookup : 

```c
Rboolean R_BindingIsActive(SEXP sym, SEXP env)
{
    if (TYPEOF(sym) != SYMSXP)
	error(_("not a symbol"));
    if (TYPEOF(env) == NILSXP)
	error(_("use of NULL environment is defunct"));
    if (TYPEOF(env) != ENVSXP &&
	TYPEOF((env = simple_as_environment(env))) != ENVSXP)
	error(_("not an environment"));
    if (env == R_BaseEnv || env == R_BaseNamespace)
	/* It is a symbol, so must have a binding even if it is
	   R_UnboundSymbol */
	return IS_ACTIVE_BINDING(sym) != 0;
    else {
	SEXP binding = findVarLocInFrame(env, sym, NULL);
	if (binding == R_NilValue)
	    error(_("no binding for \"%s\""), EncodeChar(PRINTNAME(sym)));
	return IS_ACTIVE_BINDING(binding) != 0;
    }
}
```

it feels like a waste, as we have direct access to the binding as this is traversing the frames. At least, we have tests so we could adapt if the internals were to change. 

I'll probably still change so that we don't have both : 

```cpp
bool isActiveBinding(const std::string&, const SEXP);
bool isActiveBinding(SEXP binding);
```

which feels confusing. Perhaps this one can just live in `SessionEnvironment.cpp` under a different name that makes it clearer. 

```cpp
bool isActiveBinding(SEXP binding) 
{
   static unsigned int ACTIVE_BINDING_MASK = (1<<15);
   return reinterpret_cast<r::sxpinfo*>(binding)->gp & ACTIVE_BINDING_MASK;
}
```

Reply:
Now using this locally: 

```cpp
namespace {
bool frameBindingIsActive(SEXP binding) 
{
   static unsigned int ACTIVE_BINDING_MASK = (1<<15);
   return reinterpret_cast<r::sxpinfo*>(binding)->gp & ACTIVE_BINDING_MASK;
}
}

bool frameHasExternalPointer(SEXP frame, bool nullPtr, std::set<SEXP>& visited)
{
   while(frame != R_NilValue)
   {
      if (!frameBindingIsActive(frame) && hasExternalPointer(CAR(frame), nullPtr, visited))
         return true;

      frame = CDR(frame);
   }

   return false;
}
```

This is tested